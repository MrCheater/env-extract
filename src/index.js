import JSON5 from "json5";
import crypto from "crypto";
import deepForEach from "deep-for-each";
import toDoubleQuotes from "to-double-quotes";

const regExp = /(process\.env\.(?:\w|_)(?:\w|\d|_)+?)(\s*(?:,|})(?=(?:[^"]*(?<!\\)"[^"]*(?<!\\)")*[^"]*$))/gim;

export const envKey = Symbol();

export const extractEnv = text => {
  const env = {};

  const replacer = (match, group, trail) => {
    const hash = crypto.createHash("sha256");
    hash.update(group);
    const digest = `$${group}/${hash.digest("hex")}`;
    env[digest] = group;
    return `"${digest}"${trail}`;
  };

  const json = JSON5.parse(toDoubleQuotes(text).replace(regExp, replacer));

  Object.defineProperty(json, envKey, {
    value: env
  });

  deepForEach(json, value => {
    try {
      Object.defineProperty(value, envKey, {
        value: env
      });
    } catch (e) {}
  });

  return json;
};

export const injectEnv = json => {
  let text = JSON.stringify(json, null, 2);

  const env = json[envKey];

  if (env) {
    for (const digest of Object.keys(env)) {
      while (true) {
        let prevText = text;
        text = text.replace(`"${digest}"`, env[digest]);
        if (prevText === text) {
          break;
        }
      }
    }
  }

  return text;
};
