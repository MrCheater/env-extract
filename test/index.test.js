import { extractEnv, injectEnv, envKey } from "../src";

test("works correctly", () => {
  const text = `
    {
      "abc": process.env.ABC,
      "abc2": process.env.ABC,
      'a1': process.env.A1,
      b1: process.env.B1,
      "bcd": 'process.env.BCD,',
      'a2': "process.env.A2,",
      b2: '\\'process.env.B2,',
      b3: "\\"process.env.B3,",
      b4: "\\"process.env.B4,\\"",
      json: {
        "abc": process.env.ABC
      },
      test: '""\\'"process.env.ABC,\\'"\\'"'
    }
  `;

  const json = extractEnv(text);

  expect(json).toMatchSnapshot();
  expect(json[envKey]).toMatchSnapshot();

  const result = injectEnv(json);

  expect(result).toMatchSnapshot();
});
