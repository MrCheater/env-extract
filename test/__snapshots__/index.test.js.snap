// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works correctly 1`] = `
Object {
  "a1": "$process.env.A1/658d36cb102359cca4ed408df7044db86894ae1698616794ce47c917c17c7354",
  "a2": "process.env.A2,",
  "abc": "$process.env.ABC/cdd814c371eef4408bab44025c0921610d18f14bd0df827cb68bc53318dc83d1",
  "abc2": "$process.env.ABC/cdd814c371eef4408bab44025c0921610d18f14bd0df827cb68bc53318dc83d1",
  "b1": "$process.env.B1/db25dd79c3bcb3e417489f39b20ca768dc1e14c9b2d689e445350dc1fe886a52",
  "b2": "'process.env.B2,",
  "b3": "\\"process.env.B3,",
  "b4": "\\"process.env.B4,\\"",
  "bcd": "process.env.BCD,",
  "json": Object {
    "abc": "$process.env.ABC/cdd814c371eef4408bab44025c0921610d18f14bd0df827cb68bc53318dc83d1",
    Symbol(): Object {
      "$process.env.A1/658d36cb102359cca4ed408df7044db86894ae1698616794ce47c917c17c7354": "process.env.A1",
      "$process.env.ABC/cdd814c371eef4408bab44025c0921610d18f14bd0df827cb68bc53318dc83d1": "process.env.ABC",
      "$process.env.B1/db25dd79c3bcb3e417489f39b20ca768dc1e14c9b2d689e445350dc1fe886a52": "process.env.B1",
    },
  },
  "test": "\\"\\"'\\"process.env.ABC,'\\"'\\"",
  Symbol(): Object {
    "$process.env.A1/658d36cb102359cca4ed408df7044db86894ae1698616794ce47c917c17c7354": "process.env.A1",
    "$process.env.ABC/cdd814c371eef4408bab44025c0921610d18f14bd0df827cb68bc53318dc83d1": "process.env.ABC",
    "$process.env.B1/db25dd79c3bcb3e417489f39b20ca768dc1e14c9b2d689e445350dc1fe886a52": "process.env.B1",
  },
}
`;

exports[`works correctly 2`] = `
Object {
  "$process.env.A1/658d36cb102359cca4ed408df7044db86894ae1698616794ce47c917c17c7354": "process.env.A1",
  "$process.env.ABC/cdd814c371eef4408bab44025c0921610d18f14bd0df827cb68bc53318dc83d1": "process.env.ABC",
  "$process.env.B1/db25dd79c3bcb3e417489f39b20ca768dc1e14c9b2d689e445350dc1fe886a52": "process.env.B1",
}
`;

exports[`works correctly 3`] = `
"{
  \\"abc\\": process.env.ABC,
  \\"abc2\\": process.env.ABC,
  \\"a1\\": process.env.A1,
  \\"b1\\": process.env.B1,
  \\"bcd\\": \\"process.env.BCD,\\",
  \\"a2\\": \\"process.env.A2,\\",
  \\"b2\\": \\"'process.env.B2,\\",
  \\"b3\\": \\"\\\\\\"process.env.B3,\\",
  \\"b4\\": \\"\\\\\\"process.env.B4,\\\\\\"\\",
  \\"json\\": {
    \\"abc\\": process.env.ABC
  },
  \\"test\\": \\"\\\\\\"\\\\\\"'\\\\\\"process.env.ABC,'\\\\\\"'\\\\\\"\\"
}"
`;
